pyvista/core/datasetattributes.py

```diff
@@ -166,153 +459,394 @@ def get_array(self, key: Union[str, int]) -> Union[pyvista_ndarray, _vtk.vtkData
 
-    def remove(self, key: Union[int, str]) -> None:
+        self.set_array(narray, name, deep_copy)
+        if active_scalars:
+            self.active_scalars_name = name
+
+    def remove(self, key: str) -> None:
         """Remove an array.
 
         Parameters
         ----------
-        key : int, str
-            The name or index of the array to remove.
+        key : str
+            The name of the array to remove.
+
+        Examples
+        --------
+        Add a point data array to a DataSet and then remove it.
+
+        >>> import pyvista
+        >>> mesh = pyvista.Cube().clean()
+        >>> mesh.point_arrays['my_data'] = range(mesh.n_points)
+        >>> mesh.point_arrays.remove('my_data')
+
+        Show that the array no longer exists in ``point_arrays``
+
+        >>> 'my_data' in mesh.point_arrays
+        False
+
+        Notes
+        -----
+        This is provided as VTK supports indexed arrays in DataSetAttributes.
+
         """
-        self._raise_index_out_of_bounds(index=key)
         name = self.get_array(key).GetName()  # type: ignore
         try:
-            self.dataset.association_bitarray_names[self.association].remove(name)
+            self.dataset.association_bitarray_names[self.association.name].remove(name)
         except KeyError:
             pass
         self.VTKObject.RemoveArray(key)
         self.VTKObject.Modified()
 
-    def pop(self, key: Union[int, str], default=pyvista_ndarray(array=[])) -> pyvista_ndarray:
+    def pop(self, key: str, default=pyvista_ndarray(array=[])) -> pyvista_ndarray:
         """Remove an array and return it.
 
         Parameters
         ----------
-        key : int, str
-            The name or index of the array to remove and return.
+        key : str
+            The name of the array to remove and return.
 
-        default : anything
+        default : anything, optional
             If default is not given and key is not in the dictionary,
             a KeyError is raised.
 
         Returns
         -------
-        arr : pyvista_ndarray
+        :class:`pyvista_ndarray`
             Requested array.
+
+        Examples
+        --------
+        Add a point data array to a DataSet and then remove it
+
+        >>> import pyvista
+        >>> mesh = pyvista.Cube().clean()
+        >>> mesh.point_arrays['my_data'] = range(mesh.n_points)
+        >>> mesh.point_arrays.pop('my_data')
+        pyvista_ndarray([0, 1, 2, 3, 4, 5, 6, 7])
+
+        Show that the array no longer exists in ``point_arrays``
+
+        >>> 'my_data' in mesh.point_arrays
+        False
+
         """
-        self._raise_index_out_of_bounds(index=key)
         vtk_arr = self.GetArray(key)
         if vtk_arr:
             copy = vtk_arr.NewInstance()
             copy.DeepCopy(vtk_arr)
             vtk_arr = copy
+
         try:
             self.remove(key)
         except KeyError:
             if default in self.pop.__defaults__:  # type: ignore
                 raise
```

Comment:
If I understand correctly, the mutable default argument is never actually used, it's just a weird sentinel. If this is the case then I recommend defining a `sentinel = object()` at the module global level, and then using that as the default and checking for it with `if default is sentinel: ...`.

Reply:
Quite right.  Fixed this.